src/mam/mod_mam_cache_user.erl

```diff
@@ -49,113 +48,70 @@ tbl_name_archive_id() ->
-    case gen_mod:get_module_opt(Host, ?MODULE, pm, false) of
-        true ->
-            start_pm(Host, Opts);
-        false ->
-            ok
-    end,
-    case gen_mod:get_module_opt(Host, ?MODULE, muc, false) of
-        true ->
-            start_muc(Host, Opts);
-        false ->
-            ok
-    end.
-
--spec stop(Host :: jid:server()) -> any().
-stop(Host) ->
-    stop_server(Host),
-    case gen_mod:get_module_opt(Host, ?MODULE, pm, false) of
-        true ->
-            stop_pm(Host);
-        false ->
-            ok
-    end,
-    case gen_mod:get_module_opt(Host, ?MODULE, muc, false) of
-        true ->
-            stop_muc(Host);
-        false ->
-            ok
-    end.
-
-writer_child_spec() ->
-    {?MODULE,
-     {?MODULE, start_link, []},
-     permanent,
-     5000,
-     worker,
-     [?MODULE]}.
-
-start_server(_Host) ->
-    %% TODO make per host server
-    supervisor:start_child(ejabberd_sup, writer_child_spec()).
-
-stop_server(_Host) ->
-    ok.
-
-%% ----------------------------------------------------------------------
-%% Add hooks for mod_mam
-
--spec start_pm(jid:server(), list()) -> 'ok'.
-start_pm(Host, _Opts) ->
-    ejabberd_hooks:add(mam_archive_id, Host, ?MODULE, cached_archive_id, 30),
-    ejabberd_hooks:add(mam_archive_id, Host, ?MODULE, store_archive_id, 70),
-    ejabberd_hooks:add(mam_remove_archive, Host, ?MODULE, remove_archive, 100),
+-spec start(HostType :: mongooseim:host_type(), Opts :: gen_mod:module_opts()) -> ok.
+start(HostType, _Opts) ->
+    start_server(),
+    ejabberd_hooks:add(hooks(HostType)),
     ok.
 
-
--spec stop_pm(jid:server()) -> 'ok'.
-stop_pm(Host) ->
-    ejabberd_hooks:delete(mam_archive_id, Host, ?MODULE, cached_archive_id, 30),
-    ejabberd_hooks:delete(mam_archive_id, Host, ?MODULE, store_archive_id, 70),
-    ejabberd_hooks:delete(mam_remove_archive, Host, ?MODULE, remove_archive, 100),
+-spec stop(HostType :: mongooseim:host_type()) -> ok.
+stop(HostType) ->
+    ejabberd_hooks:delete(hooks(HostType)),
     ok.
 
+writer_child_spec() ->
+    MFA = {?MODULE, start_link, []},
+    {?MODULE, MFA, permanent, 5000, worker, [?MODULE]}.
 
-%% ----------------------------------------------------------------------
-%% Add hooks for mod_mam_muc
+start_server() ->
+    %% TODO make per host server
+    supervisor:start_child(ejabberd_sup, writer_child_spec()).
 
--spec start_muc(jid:server(), list()) -> 'ok'.
-start_muc(Host, _Opts) ->
-    ejabberd_hooks:add(mam_muc_archive_id, Host, ?MODULE, cached_archive_id, 30),
-    ejabberd_hooks:add(mam_muc_archive_id, Host, ?MODULE, store_archive_id, 70),
-    ejabberd_hooks:add(mam_muc_remove_archive, Host, ?MODULE, remove_archive, 100),
-    ok.
+hooks(HostType) ->
+    PM = gen_mod:get_module_opt(HostType, ?MODULE, pm, false),
+    MUC = gen_mod:get_module_opt(HostType, ?MODULE, muc, false),
+    case PM of
+        true ->
+            pm_hooks(HostType);
+        false ->
+            []
+    end ++
+    case MUC of
```

Comment:
It's good, the only problem I see is that the part after `++` should be indented and with correct indentation it looks worse.